Flaky TestDistributor_LabelValuesCardinality

```
--- FAIL: TestDistributor_LabelValuesCardinality (0.00s)
    --- FAIL: TestDistributor_LabelValuesCardinality/should_return_a_map_with_the_label_values_and_series_occurrences_of_a_single_label_name (0.00s)
        --- FAIL: TestDistributor_LabelValuesCardinality/should_return_a_map_with_the_label_values_and_series_occurrences_of_a_single_label_name/minimize_ingester_requests:_true (0.17s)
            distributor_test.go:4595: 
                	Error Trace:	/__w/mimir/mimir/pkg/distributor/distributor_test.go:4595
                	            				/__w/mimir/mimir/vendor/github.com/grafana/dskit/test/poll.go:17
                	            				/__w/mimir/mimir/pkg/distributor/distributor_test.go:4592
                	Error:      	Not equal: 
                	            	expected: 0x3
                	            	actual  : 0x2
                	Test:       	TestDistributor_LabelValuesCardinality/should_return_a_map_with_the_label_values_and_series_occurrences_of_a_single_label_name/minimize_ingester_requests:_true
```

Failed for me today too, plus:

```
$ go test -failfast -count=100 -run=TestDistributor_LabelValuesCardinality ./pkg/distributor
--- FAIL: TestDistributor_LabelValuesCardinality (0.00s)
    --- FAIL: TestDistributor_LabelValuesCardinality/should_return_a_map_with_the_label_values_and_series_occurrences_of_all_the_label_names (0.00s)
        --- FAIL: TestDistributor_LabelValuesCardinality/should_return_a_map_with_the_label_values_and_series_occurrences_of_all_the_label_names/minimize_ingester_requests:_false (0.12s)
            distributor_test.go:4588:
                	Error Trace:	/Users/oleg/w/github.com/grafana/mimir/pkg/distributor/distributor_test.go:4588
                	            				/Users/oleg/w/github.com/grafana/mimir/vendor/github.com/grafana/dskit/test/poll.go:17
                	            				/Users/oleg/w/github.com/grafana/mimir/pkg/distributor/distributor_test.go:4585
                	Error:      	Not equal:
                	            	expected: 0x3
                	            	actual  : 0x2
                	Test:       	TestDistributor_LabelValuesCardinality/should_return_a_map_with_the_label_values_and_series_occurrences_of_all_the_label_names/minimize_ingester_requests:_false
FAIL
FAIL	github.com/grafana/mimir/pkg/distributor	17.091s
FAIL
```

--- a/pkg/distributor/distributor_test.go
+++ b/pkg/distributor/distributor_test.go
@@ -4583,16 +4583,19 @@ func TestDistributor_LabelValuesCardinality(t *testing.T) {
 					// Since the Push() response is sent as soon as the quorum is reached, when we reach this point
 					// the final ingester may not have received series yet.
 					// To avoid flaky test we retry the assertions until we hit the desired state within a reasonable timeout.
-					test.Poll(t, time.Second, testData.expectedResult, func() interface{} {
+					require.Eventually(t, func() bool {
 						seriesCountTotal, cardinalityMap, err := ds[0].LabelValuesCardinality(ctx, testData.labelNames, testData.matchers, cardinality.InMemoryMethod)
-						require.NoError(t, err)
-						assert.Equal(t, testData.expectedSeriesCountTotal, seriesCountTotal)
-						// Make sure the resultant label names are sorted
+						if !assert.NoError(t, err) {
+							return false
+						}
+						if !assert.Equal(t, testData.expectedSeriesCountTotal, seriesCountTotal) {
+							return false
+						}
 						sort.Slice(cardinalityMap.Items, func(l, r int) bool {
 							return cardinalityMap.Items[l].LabelName < cardinalityMap.Items[r].LabelName
 						})
-						return cardinalityMap
-					})
+						return assert.EqualValues(t, testData.expectedResult, cardinalityMap)
+					}, 1*time.Second, 50*time.Millisecond)
 
 					// Make sure enough ingesters were queried
 					assert.GreaterOrEqual(t, countMockIngestersCalled(ingesters, "LabelValuesCardinality"), testData.expectedIngesters)
@@ -5669,7 +5672,6 @@ func prepareIngesters(t testing.TB, cfg prepConfig) []*mockIngester {
 		ingesters = append(ingesters, prepareIngesterZone(t, zone, state, cfg)...)
 	}
 	return ingesters
-
 }
 
 func prepareIngesterZone(t testing.TB, zone string, state ingesterZoneState, cfg prepConfig) []*mockIngester {

--- a/pkg/distributor/distributor_test.go
+++ b/pkg/distributor/distributor_test.go
@@ -4585,12 +4585,14 @@ func TestDistributor_LabelValuesCardinality(t *testing.T) {
 					// To avoid flaky test we retry the assertions until we hit the desired state within a reasonable timeout.
 					require.Eventually(t, func() bool {
 						seriesCountTotal, cardinalityMap, err := ds[0].LabelValuesCardinality(ctx, testData.labelNames, testData.matchers, cardinality.InMemoryMethod)
-						if !assert.NoError(t, err) {
+						if err != nil {
 							return false
 						}
-						if !assert.Equal(t, testData.expectedSeriesCountTotal, seriesCountTotal) {
+						if testData.expectedSeriesCountTotal != seriesCountTotal {
+							// Not all ingesters have received the series yet.
 							return false
 						}
+
 						sort.Slice(cardinalityMap.Items, func(l, r int) bool {
 							return cardinalityMap.Items[l].LabelName < cardinalityMap.Items[r].LabelName
 						})

